File: spark/examples/s3-select.py
'''Use spark-select to read from an S3 instance.'''

import operator
import pyspark
import os
import sys
from pyspark.sql import SparkSession
from pyspark.sql.types import *

def main(s3_ip_addr):
    #os.environ['PYSPARK_SUBMIT_ARGS'] = "--packages=org.apache.hadoop:hadoop-aws:2.7.3 pyspark-shell"

    #Intialize a spark context
    with pyspark.SparkContext("local", "PySparkWordCount") as sc:

        sc.setLogLevel('WARN')
        hadoopConf = sc._jsc.hadoopConfiguration()
        hadoopConf.set("fs.s3a.access.key", "admin")
        hadoopConf.set("fs.s3a.secret.key", "admin123")
        hadoopConf.set("fs.s3a.endpoint", "http://{}:9000".format(s3_ip_addr))
        hadoopConf.set("fs.s3a.path.style.access", "true")
        hadoopConf.set("fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")

        spark = SparkSession.builder \
                        .appName("Python Spark SQL basic example") \
                        .config("spark.some.config.option", "some-value") \
                        .getOrCreate()

        # filtered schema
        st = StructType([
            StructField("id", IntegerType(), False),
            StructField("name", StringType(), True),
            StructField("age", IntegerType(), False),
            StructField("city", StringType(), True),
        ])

        df = spark \
            .read \
            .format('minioSelectCSV') \
            .schema(st) \
            .load("s3a://spark-test/s3_data.csv")

        # show all rows.
        df.show()

        # show only filtered rows.
        df.select("*").filter("age > 40").show()
        #df.select(df['name'],df['age'],df['id']).filter(df.name.like("jim")).filter("age > 40").show()
        df.select(df['name'],df['age'],df['id']).filter(df.name.like('Jim%')).filter("age > 40").show()
        df.select(df['name'],df['age'],df['id']).filter(df.city.like('Miami%')).filter(df.name.like('Jim%')).filter("age > 40").show()
        df.select(df['name'],df['age'],df['id']).filter("age < 90").filter("age > 40").show()
        df.select(df['name'],df['age'],df['id']).filter("id < 5").filter("age > 40").show()
    

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("ERROR: missing ip addr of s3 instance")
        print("s3: <ip addr of s3>")
        exit(1)
    ip_addr = sys.argv[1]
    main(ip_addr)

File: spark/examples/scala/src/main/scala/s3-select.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.types._

object SparkTest {
    def main(args: Array[String]) = {

        if (args.length == 0) {
            println("missing arg for s3 ip addr") 
            System.exit(1)
        }
        val s3IpAddr = args(0)
        val spark: SparkSession = SparkSession.builder()
          .master("local[1]")
          .appName("s3-select example")
          .getOrCreate()
        spark.sparkContext.setLogLevel("ERROR") // WARN INFO TRACE DEBUG
        spark.sparkContext.hadoopConfiguration.set("fs.s3a.access.key", "admin")
        spark.sparkContext.hadoopConfiguration.set("fs.s3a.secret.key", "admin123")
        spark.sparkContext.hadoopConfiguration.set("fs.s3a.endpoint", 
                                                   s"""http://$s3IpAddr:9000""")
        spark.sparkContext.hadoopConfiguration.set("fs.s3a.path.style.access", "true")
        spark.sparkContext.hadoopConfiguration.set("fs.s3a.impl", 
                                           "org.apache.hadoop.fs.s3a.S3AFileSystem")
        // filtered schema
        val schema = new StructType()
            .add("id",IntegerType,true)
            .add("name",StringType,true)
            .add("age",IntegerType,true)
            .add("city",StringType,true)

        var df = spark
            .read
            .format("minioSelectCSV")
            .schema(schema)
            .load("s3a://spark-test/s3_data.csv")

        // show all rows.
        df.show()

        // show only filtered rows.
        df.select("*").filter("age > 40").show()
        //df.select(df("name"),df("age"),df("id")).filter(df.name.like("jim")).filter("age > 40").show()
        //df.select(df("name"),df("age"),df("id")).filter(df.name.like("Jim%")).filter("age > 40").show()
        df.select(df("name"),df("age"),df("id"))
                 .filter(df.col("city").like("Miami%")).filter(df.col("name").like("Jim%")).filter("age > 40").show()
        df.select(df("name"),df("age"),df("id")).filter("age < 90").filter("age > 40").show()
        df.select(df("name"),df("age"),df("id")).filter("id < 5").filter("age > 40").show()
        var c = df.select(df("name"),df("age"),df("id")).filter("id < 5").filter("age > 40").count()
        println(s"""count is $c""")
        println("S3 Test Successful")
    }
}
